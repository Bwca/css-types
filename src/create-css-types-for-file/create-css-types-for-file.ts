import { writeFileSync } from 'fs';

import { EnumGenerator } from '../enum-generator/enum-generator';
import { extractCssStringFromFile } from '../extract-css-string-from-file/extract-css-string-from-file';
import { generateDefinitionsFileName } from '../generate-definitions-file-name/generate-definitions-file-name';
import { getCssClassesAsArray } from '../get-compiled-classes/get-compiled-classes';
import { NotificationService } from '../notification-service';

export async function createCssTypesForFile(filePath: string): Promise<void> {
  const cssString = await extractCssStringFromFile(filePath);

  if (!cssString) {
    NotificationService.warning(`Could not obtain ${filePath} css from file!`);
    return;
  }

  const cssClasses = getCssClassesAsArray(cssString);

  if (!cssClasses) {
    NotificationService.error(`Could not parse css classes from ${filePath}`);
    return;
  }

  const enumDefinition = EnumGenerator.generateEnumString(filePath, cssClasses);
  const definitionsFileName = generateDefinitionsFileName(filePath);

  try {
    writeFileSync(definitionsFileName, enumDefinition, {
      encoding: 'utf-8',
      flag: 'w+',
    });

    NotificationService.success(`Successfully wrote ${cssClasses.length} keys length enum from ${filePath}`);
  } catch (e) {
    NotificationService.error(`Error writing types enum for ${filePath}`);
  }
}
