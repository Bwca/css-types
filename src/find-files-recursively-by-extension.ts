import { readdirSync, statSync } from 'fs';
import { join } from 'path';

export function findFilesRecursivelyByExtension(dir: string, extension = /scss$/, files: string[] = []): string[] {
  return (
    readdirSync(dir)
      .map((potentialFile) => {
        const dirPath = join(dir, potentialFile);
        const isDirectory = statSync(dirPath).isDirectory();
        if (isDirectory) {
          return findFilesRecursivelyByExtension(dirPath, extension, files);
        }
        if (extension.test(potentialFile)) {
          return [...files, join(dir, potentialFile)];
        }
        return null;
      })
      .filter(Boolean)
      .reduce((a, b) => [...a, ...b], []) || []
  );
}
