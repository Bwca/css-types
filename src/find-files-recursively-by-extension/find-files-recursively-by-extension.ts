import { readdirSync, statSync } from 'fs';
import { join } from 'path';

export function findFilesRecursivelyByExtension(dir: string, extension = /scss$/, files: string[] = []): string[] {
  return (
    readdirSync(dir)
      .map((potentialFile) => {
        const dirPath = join(dir, potentialFile);
        const isDirectory = statSync(dirPath).isDirectory();
        if (isDirectory) {
          return findFilesRecursivelyByExtension(dirPath, extension, files);
        }
        if (extension.test(potentialFile)) {
          return [...files, join(dir, potentialFile)];
        }
      })
      .filter(Boolean)
      .reduce((a, b) => [...(a as string[]), ...(b as string[])], []) || []
  );
}
