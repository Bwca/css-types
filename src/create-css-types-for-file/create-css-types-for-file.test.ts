import { existsSync, readFileSync, unlinkSync } from 'fs';

import { findFilesRecursivelyByExtension } from '../find-files-recursively-by-extension/find-files-recursively-by-extension';
import { createCssTypesForFile } from './create-css-types-for-file';

describe(`${createCssTypesForFile.name} test`, () => {
  /** Cleanup any existing enum files before tests */
  beforeAll(() => {
    const testDir = __dirname + '/__test__';
    const existingTypescriptEnums = findFilesRecursivelyByExtension(testDir, /\.ts$/, []);
    existingTypescriptEnums.forEach((f) => unlinkSync(f));
  });

  it('should create types file for a css stylesheet', async () => {
    // Arrange
    const stylesheet = __dirname + '/__test__/css/sheet.css';
    const expectedEnumFile = __dirname + '/__test__/css/sheet.style.ts';

    // Act
    await createCssTypesForFile(stylesheet);

    // Assert
    expect(existsSync(expectedEnumFile)).toBeTruthy();
    expect(readFileSync(expectedEnumFile, 'utf8')).toMatchSnapshot();
  });

  it('should create types file for scss stylesheet', async () => {
    // Arrange
    const stylesheet = __dirname + '/__test__/scss/sheet.scss';
    const expectedEnumFile = __dirname + '/__test__/scss/sheet.style.ts';

    // Act
    await createCssTypesForFile(stylesheet);

    // Assert
    expect(existsSync(expectedEnumFile)).toBeTruthy();
    expect(readFileSync(expectedEnumFile, 'utf8')).toMatchSnapshot();
  });
});
