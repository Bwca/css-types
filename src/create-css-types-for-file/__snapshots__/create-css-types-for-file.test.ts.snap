// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCssTypesForFile test should create types file for a css stylesheet 1`] = `
"export enum Sheet {
  One = 'one',
  Two = 'two',
  Three = 'three',
  Four = 'four',
  Five = 'five',
}
"
`;

exports[`createCssTypesForFile test should create types file for scss stylesheet 1`] = `
"export enum Sheet {
  One = 'one',
  Two = 'two',
  Three = 'three',
  Five = 'five',
  Four = 'four',
}
"
`;
